solr/core/src/java/org/apache/solr/servlet/CoreContainerProvider.java

```diff
@@ -93,69 +87,71 @@ public class CoreContainerProvider implements ServletContextListener {
   private HttpClient httpClient;
   private SolrMetricManager metricManager;
   private RateLimitManager rateLimitManager;
-  private final CountDownLatch init = new CountDownLatch(1);
   private String registryName;
-  // AFAIK the only reason we need this is to support JettySolrRunner for tests. In tests we might
-  // have multiple CoreContainers in the same JVM, but I *think* that doesn't happen in a real
-  // server.
-  private static final Map<ContextInitializationKey, ServiceHolder> services =
-      Collections.synchronizedMap(new WeakHashMap<>());
-
-  // todo: dependency injection instead, but for now this method and the associated map will have
-  // to suffice.
-  // Note that this relies on ServletContext.equals() not implementing anything significantly
-  // different than Object.equals for its .equals method (I've found no implementation that even
-  // implements it).
-  public static ServiceHolder serviceForContext(ServletContext ctx) throws InterruptedException {
-    ContextInitializationKey key = new ContextInitializationKey(ctx);
-    return services.computeIfAbsent(key, ServiceHolder::new);
+
+  /** Acquires an instance from the context, waiting if necessary. */
+  public static CoreContainerProvider serviceForContext(ServletContext ctx)
+      throws InterruptedException {
+    long startWait = System.nanoTime();
+    synchronized (ctx) {
+      CoreContainerProvider provider;
+      while ((provider =
+              (CoreContainerProvider) ctx.getAttribute(CoreContainerProvider.class.getName()))
+          == null) {
+        ctx.wait(1_000);
+        final int seconds = (int) ((System.nanoTime() - startWait) / 1_000_000_000);
+        if (seconds >= 2 && log.isInfoEnabled()) {
+          log.info("Still waiting for CoreContainer startup ({} seconds elapsed)", seconds);
+        }
+      }
+      return provider;
+    }
   }
 
   @Override
-  public void contextInitialized(ServletContextEvent sce) {
-    init(sce.getServletContext());
+  public void contextInitialized(ServletContextEvent event) {
+    final var ctx = event.getServletContext();
+    init(ctx);
+    ctx.setAttribute(CoreContainerProvider.class.getName(), this);
+    // let serviceForContext() know
+    synchronized (ctx) {
+      ctx.notifyAll();
+    }
   }
 
   @Override
   public void contextDestroyed(ServletContextEvent sce) {
     close();
   }
 
+  /**
+   * @see SolrDispatchFilter#getCores()
+   */
   CoreContainer getCoreContainer() throws UnavailableException {
-    waitForCoreContainer(() -> cores, init);
```

Comment:
Don't wait at this point anymore; it ought to be ready (so we check to report a nice exception if it isn't)